packages/material-ui/src/TextField/TextField.js

```diff
@@ -54,133 +53,123 @@ const styles = {
  * - using the upper case props for passing values directly to the components
  * - using the underlying components directly as shown in the demos
  */
-class TextField extends React.Component {
-  constructor(props) {
-    super(props);
-    this.labelRef = React.createRef();
-  }
+const TextField = React.forwardRef(function TextField(props, ref) {
+  const {
+    autoComplete,
+    autoFocus,
+    children,
+    classes,
+    className: classNameProp,
+    defaultValue,
+    error,
+    FormHelperTextProps,
+    fullWidth,
+    helperText,
+    id,
+    InputLabelProps,
+    inputProps,
+    InputProps,
+    inputRef,
+    label,
+    multiline,
+    name,
+    onBlur,
+    onChange,
+    onFocus,
+    placeholder,
+    required,
+    rows,
+    rowsMax,
+    select,
+    SelectProps,
+    type,
+    value,
+    variant,
+    ...other
+  } = props;
 
-  componentDidMount() {
-    if (this.props.variant === 'outlined') {
-      this.labelNode = ReactDOM.findDOMNode(this.labelRef.current);
-      this.forceUpdate();
+  const [labelWidth, setLabelWidth] = React.useState(0);
+  const labelRef = React.useRef();
+  React.useEffect(() => {
+    if (variant === 'outlined') {
+      // StrictMode ready
```

Comment:
I'll be annotating every `findDOMNode` usage that calls refs with this comment. Can't fully switch in case people pass a class to `InputLabel` resulting in `<InputLabel component={ClassComponent} />`.

Reply:
I would like to re-architecture this logic. There are two problems with it.

1. People not using the TextField components have to repeat it, we could provide the reference instead of the actual width, let the OutlinedInput do the necessary work.
1. Because the OutlinedInput doesn't have access to the reference, he can't update the width value when something changes.

Reply:
> the reference instead of the actual width, let the OutlinedInput do the necessary work.

How would that work look like? When should the OutlinedInput update the width?

Reply:
It's a great question. I don't have any good answer to it. Let's keep it this way for now.